spmd/tensor/placement_types.py

```diff
@@ -30,6 +32,170 @@ class Shard(Placement):
     # shard placement, shard on a dim
     dim: int
 
+    def _split_tensor(
+        self,
+        tensor: torch.Tensor,
+        num_chunks: int,
+        *,
+        with_padding: bool = True,
+        contiguous: bool = True,
+    ) -> Tuple[List[torch.Tensor], int]:
+        # NOTE: For with_padding option, we pad the tensor on each rank before calling
+        # the collectives (i.e. scatter/all_gather, etc.). This is because for gloo
+        # backend, it does not support uneven collectives, nccl supports some, but
+        # it might be slow compared to even size collective, we need to pad tensor
+        # before really calling the collective, and unpad/narrow it afterwards
+        # TODO: consider if we should remove this logic once ProcessGroupGloo
+        # support uneven list, and collective perfomance on par
+        assert (
+            self.dim <= tensor.ndim
+        ), f"Sharding dim {self.dim} greater than tensor ndim {tensor.ndim}"
+        assert (
+            tensor.size(self.dim) >= num_chunks
+        ), f"Tensors to be sharded on dim {self.dim} must be at least as large as the number of devices in that dimension {num_chunks}"
```

Comment:
Is this assertion necessary now that we support uneven shards? E.g. if we have 4 devices but a tensor of size 3, we could support it by having a device holding an empty tensor (or a 1 element tensor in case with_padding=True).

Reply:
Got it so you did try it out and it didn't work? (I know that c10d segfaults when passed an empty tensor ... but here we're padding, so that shouldn't be a problem).